### PR TITLE
tech (css-library): replace ui-library with twbs 5.0

### PR DESCRIPTION
**Pull request by cerebrl**
**Summary:**

The internally used UI style library is no longer maintained, so this replaces it by directly using Twitter Bootstrap v5.

**Breaking Changes:**

Some markup and class changes occurred within the password and select components.

**Demo:**

https://share.getcloudapp.com/4gunYzpm

### CODE REVIEW

File: samples/oauth/index.html [hunk 1]
```diff
@@ -59,15 +59,16 @@
...
-        clientId: '<Your OAuth Client>', // string; e.g. 'ForgeRockSDKClient'
-        redirectUri: '<Your Redirect/Callback URL>', // string; e.g. 'https://sdkapp.example.com:8443/_callback'
-        scope: '<Your OAuth Scopes>', // string; e.g. 'openid profile me.read'
+        clientId: 'WebOAuthClient', // e.g. 'ForgeRockSDKClient'
+        redirectUri: 'https://sdkapp.example.com:8443/_callback/', // e.g. 'https://sdkapp.example.com:8443/central-login/'
+        scope: 'openid profile me.read', // e.g. 'openid profile me.read'
         serverConfig: {
-          baseUrl: '<Your AM URL>', // string; e.g. 'https://openam.example.com:9443/openam/'
-          timeout: '<Your Timeout (milliseconds)>' // number; 90000 or less
+          baseUrl: 'https://openam-crbrl-01.forgeblocks.com/am/', // e.g. 'https://openam.example.com:9443/openam/'
```
Comment (george-bafaloukas-forgerock) [inline samples/oauth/index.html:66]: I think is better to sanitise those and not have your env conf in the PR

File: src/basic/renderers/choice.ts [hunk 2]
```diff
@@ -78,6 +69,16 @@ class ChoiceCallbackRenderer implements DestroyableCallbackRenderer, FocusableCa
...
+    // Add the prompt
+    const prompt = this.callback.getPrompt();
+    if (prompt) {
+      formGroup.classList.add('form-floating');
+      const label = el<HTMLLabelElement>('label');
+      label.innerHTML = prompt;
+      label.htmlFor = id;
+      formGroup.appendChild(label);
+    }
+
```
Comment (cerebrl) [inline src/basic/renderers/choice.ts:81]: This was moved down from above, the only new code was the `formGroup.classList.add()` for adding the appropriate CSS class.

File: src/basic/renderers/choice.ts [hunk 3]
```diff
@@ -53,20 +53,11 @@ class ChoiceCallbackRenderer implements DestroyableCallbackRenderer, FocusableCa
...
-    if (prompt) {
-      const label = el<HTMLLabelElement>('label');
-      label.innerHTML = prompt;
-      label.htmlFor = id;
-      formGroup.appendChild(label);
-    }
-
     // Add the dropdown
-    this.input = el<HTMLSelectElement>('select', 'form-control');
+    this.input = el<HTMLSelectElement>('select', 'form-select');
```
Comment (cerebrl) [inline src/basic/renderers/choice.ts:60]: Needed to change this to provide the appropriate styling without `ui-design` library.

File: src/basic/renderers/choice.ts [hunk 4]
```diff
@@ -53,20 +53,11 @@ class ChoiceCallbackRenderer implements DestroyableCallbackRenderer, FocusableCa
    */
   public render = (): HTMLDivElement => {
     const defaultIndex = this.callback.getInputValue() as number;
-    const formGroup = el<HTMLDivElement>('div', `fr-callback-${this.index} form-group`);
+    const formGroup = el<HTMLDivElement>('div', `fr-callback-${this.index} form-group mb-3`);
```
Comment (cerebrl) [inline src/basic/renderers/choice.ts:56]: `mb-3` just means "margin-bottom" for 3 units of an arbitrary amount.

File: src/basic/renderers/kba-create.ts [hunk 5]
```diff
@@ -96,22 +97,31 @@ class KbaCreateCallbackRenderer implements DestroyableCallbackRenderer, Focusabl
     const selectedIndex = allOptions.indexOf(isCustomQuestion ? TEXT.customLabel : questionText);
 
     // Add the dropdown
-    const selectWrap = el<HTMLDivElement>('div', 'mb-1');
-    this.select = el<HTMLSelectElement>('select', 'form-control custom-select');
+    const selectWrap = el<HTMLDivElement>('div', 'mb-1 form-floating');
+    this.select = el<HTMLSelectElement>('select', 'form-select');
```
Comment (cerebrl) [inline src/basic/renderers/kba-create.ts:101]: Changed to `form-select` for appropriate styling.

File: src/basic/renderers/kba-create.ts [hunk 6]
```diff
@@ -96,22 +97,31 @@ class KbaCreateCallbackRenderer implements DestroyableCallbackRenderer, Focusabl
...
+    this.select.id = `fr-callback-${this.index}-question`;
```
Comment (cerebrl) [inline src/basic/renderers/kba-create.ts:102]: Added for better DOM accessibility.

File: src/basic/renderers/kba-create.ts [hunk 7]
```diff
@@ -96,22 +97,31 @@ class KbaCreateCallbackRenderer implements DestroyableCallbackRenderer, Focusabl
...
     allOptions.forEach((x, i) => {
       const option = this.createOption(x, i === selectedIndex);
       this.select.appendChild(option);
     });
     this.select.addEventListener('change', this.onQuestionChange);
     selectWrap.appendChild(this.select);
+    const selectLabel = el<HTMLLabelElement>('label');
+    selectLabel.htmlFor = this.select.id;
+    selectLabel.innerHTML = TEXT.questionLabel;
+    selectWrap.appendChild(selectLabel);
```
Comment (cerebrl) [inline src/basic/renderers/kba-create.ts:113]: Added for "floating-label" style.

File: src/basic/renderers/kba-create.ts [hunk 8]
```diff
@@ -96,22 +97,31 @@ class KbaCreateCallbackRenderer implements DestroyableCallbackRenderer, Focusabl
...
     formGroup.appendChild(selectWrap);
 
     // Create the "custom question" input
-    this.customWrap = el<HTMLDivElement>('div', 'form-label-group fr-kba-custom-wrap mb-1 d-none');
+    this.customWrap = el<HTMLDivElement>(
+      'div',
+      'form-label-group form-floating fr-kba-custom-wrap mb-1 d-none',
```
Comment (cerebrl) [inline src/basic/renderers/kba-create.ts:119]: Just added a new class.

File: src/basic/renderers/password.ts [hunk 9]
```diff
@@ -24,7 +24,6 @@ import { renderErrors } from './errors';
 class PasswordCallbackRenderer implements DestroyableCallbackRenderer, FocusableCallbackRenderer {
   private input!: HTMLInputElement;
   private toggle!: HTMLButtonElement;
-  private toggleIcon!: HTMLElement;
```
Comment (cerebrl) [inline src/basic/renderers/password.ts:27]: No longer using the old icon from the UI design library.

File: src/basic/renderers/password.ts [hunk 10]
```diff
@@ -60,19 +59,18 @@ class PasswordCallbackRenderer implements DestroyableCallbackRenderer, Focusable
...
-    const formLabelGroup = el<HTMLDivElement>('div', 'form-label-group form-label-group-password');
-    const formLabelGroupInput = el<HTMLDivElement>('div', 'form-label-group-input');
-    const formInputGroupAppend = el<HTMLDivElement>('div', 'input-group-append');
-    formLabelGroup.appendChild(formLabelGroupInput);
-    formLabelGroup.appendChild(formInputGroupAppend);
+    const formGroup = el<HTMLDivElement>('div', `fr-callback-${this.index} form-group mb-3`);
+    const formLabelGroup = el<HTMLDivElement>(
+      'div',
+      'form-floating form-label-group input-group form-label-group-password',
+    );
```
Comment (cerebrl) [inline src/basic/renderers/password.ts:66]: This was the only component that required the most refactoring. It's mostly for the "show password" feature to switch between plain text and masked.

File: src/css/sdk.scss [hunk 11]
```diff
@@ -1,4 +1,20 @@
...
+ * sdk.scss
+ *
+ * Copyright (c) 2020-2021 ForgeRock. All rights reserved.
+ * This software may be modified and distributed under the terms
+ * of the MIT license. See the LICENSE file for details.
+ */
+ /**
+ * Import Twitter Bootstrap and its overrides from `ui-design.scss`
+ */
 @import 'ui-design';
```
Comment (cerebrl) [inline src/css/sdk.scss:13]: This is a reference to a file named "ui-design" and _not_ the library that was removed. To reduce risk, I'm trying to keep as much of the code and file names the same as before.

File: src/css/ui-design.scss [hunk 12]
```diff
@@ -1,29 +1,31 @@
...
-@import 'placeholders';
-@import 'position';
-@import 'select';
-@import 'sidebar';
-@import 'spinners';
-@import 'sprite';
-@import 'switch';
-@import 'tables';
-@import 'tabs';
-@import 'typography';
```
Comment (cerebrl) [inline src/css/ui-design.scss:29]: These were all the ui-design library components. The Bootstrap components are now being used directly.

File: webpack.config.scss.js [hunk 13]
```diff
@@ -42,11 +42,30 @@ module.exports = (env) => {
                 name: 'fr-ui.css',
               },
             },
+            {
+              // Run postcss actions
+              loader: 'postcss-loader',
```
Comment (cerebrl) [inline webpack.config.scss.js:47]: Adds better browser compatibility when CSS is compiled.

File: samples/oauth/index.html (hunk 1 above)
Comment (cerebrl) [reply samples/oauth/index.html]: Fixed, thanks for the catch!

**Review by spetrov — COMMENTED** [review]
Looks good to me! I did test various nodes (callbacks) and except the side-effect that happens with the device collector node everything else is looking good!

**Review by spetrov — APPROVED** [review]

**Pull request merged**